unitTests/ui/QuerySuggestPreviewTest.ts

```diff
@@ -231,5 +287,54 @@ export function QuerySuggestPreviewTest() {
         }, test.cmp.options.executeQueryDelay);
       });
     });
+    describe('Analytics', () => {
+      function getAResult(done) {
+        const previewContainer = $$(suggestionContainer.el).find('.coveo-preview-results > .CoveoResult');
+        if (!previewContainer) {
+          done.fail('No result to click. Impossible validate the analytics');
+        }
+        return previewContainer;
+      }
+
+      it('it log an analytics with the appropriate event', done => {
+        const suggestion = 'test';
+        triggerQuerySuggestHover(suggestion);
+        setTimeout(() => {
+          const previewContainer = getAResult(done);
+          previewContainer.click();
+          expect(test.cmp.usageAnalytics.logCustomEvent).toHaveBeenCalledWith(
+            analyticsActionCauseList.clickQuerySuggestPreview,
+            jasmine.objectContaining({
+              suggestion,
+              displayedRank: 0
+            }),
+            previewContainer
+          );
+          done();
+        }, test.cmp.options.executeQueryDelay);
+      });
+
+      it(`it log an analytics with the appropriate event,
+      even if we hover on another suggestion before clicking`, done => {
+        const suggestion = 'test';
+        triggerQuerySuggestHover(suggestion);
+        setTimeout(() => {
```

Comment:
Would jasmine.clock work instead of the setTimeout? It would help a lot with the readability of the test by reducing the nesting.

Reply:
The `jasmine.clock` and `Promise` really don't work well together and what i worse is when we don't directly create the promise so we can't really wait for it.

While browsing issue on the their github, i found a trick that people use to reduce the nesting. Here's what it'll look like.


```
function waitXms(ms: number) {
  return Promise.resolve(setTimeout(() => {return;},ms))
}

it(`it log an analytics with the appropriate event,
even if we hover on another suggestion before clicking`, async done => {
  const suggestion = 'test';
  triggerQuerySuggestHover(suggestion);
  await waitXms(test.cmp.options.executeQueryDelay)
  triggerQuerySuggestHover(`bad ${suggestion}`);
  await waitXms(test.cmp.options.executeQueryDelay - 100)
  const previewContainer = getAResult(done);
  previewContainer.click();
  expect(test.cmp.usageAnalytics.logCustomEvent).toHaveBeenCalledWith(
   analyticsActionCauseList.clickQuerySuggestPreview,
   jasmine.objectContaining({
     suggestion,
     displayedRank: 0
     }),
     previewContainer
  );
  done();
});
```

I will use this function everywhere I use the setTimeout since it is a lot cleaner.